### ISSUE TITLE
custom route constraint not called unless endpoint routing is disabled

### ISSUE THREAD

**Issue by joeaudette** [aspnet/AspNetCore#4591]
this is possibly related to #4414 

I'm having a similar problem where my custom route constraint is not being called unless I disable endpoint routing.
My constraint is for multi-tenancy based on the first folder segment of the url. If the first segment maps to a tenant then that tenant is resolved otherwise the default tenant is resolved. The route constaint is intended to keep the folder segment for other urls and links.

My constraint looks like this:

    public class SiteFolderRouteConstraint : IRouteConstraint
    {
        public SiteFolderRouteConstraint()
        {
            
        }

        public bool Match(
            HttpContext httpContext, 
            IRouter route, 
            string routeKey, 
            RouteValueDictionary values, 
            RouteDirection routeDirection)
        {
            string requestFolder = httpContext.Request.Path.ToString().StartingSegment();
            var tenant = httpContext.GetTenant<SiteContext>();
            if (tenant != null)
            {
                if (tenant.SiteFolderName == requestFolder) { return true; }

            }
            return false;
        }
    }

and is registered like this:

            services.AddRouting(options =>
            {
                options.ConstraintMap.Add("sitefolder", typeof(SiteFolderRouteConstraint));
            });

when I set a breakpoint in my constraint it is not hit unless I disable endpoint routing.

So for example my default tenant is at the root https://localhost:44399 and I have a second tenant at https://localhost:44399/two

I have a default route for folder tenants like this:

    routes.MapRoute(
                    name: "folderdefault",
                    template: "{sitefolder}/{controller}/{action}/{id?}",
                    defaults: new { controller = "Home", action = "Index" },
                    constraints: new { name = new cloudscribe.Core.Web.Components.SiteFolderRouteConstraint() });

followed by a default rout for the root tenant:

    routes.MapRoute(
                name: "default",
                template: "{controller}/{action}"
                ,defaults: new { controller = "Home", action = "Index" }
                );

I have links that are resolved using just controller and action, but in 2.1 the links would include the folder segment if the current page was already using the folder segment, ie if I',m on the home page of the folder tenant https://localhost:44399/two then the login link would be https://localhost:44399/two/account/login whereas on the root tenant home page it would be resolved as https://localhost:44399/account/login

But with endpoint routing enabled the login link is not resolved correctly for the folder tenant and no breakpoint is hit in my route constraint.

**Comment by JamesNK** [comment]
I've tracked this down to how MVC uses endpoint routing in 2.2. When the MVC endpoint datasource denormalizes conventional routes it won't use constraints that are not associated with a parameter.

For example, the constraint "name" does not have a matching route parameter and the constraint is not used by endpoint routing:
```
routes.MapRoute(
                name: "folderdefault",
                template: "{sitefolder}/{controller}/{action}/{id?}",
                defaults: new { controller = "Home", action = "Index" },
                constraints: new { name = new cloudscribe.Core.Web.Components.SiteFolderRouteConstraint() });
```

This issue is already fixed in 3.0. The MVC endpoint datasource was dramatically simplified and routes are no longer denormalized in MVC.

The workaround to this issue is to set `EnabledEndpointRouting` to false.

Thoughts on fixing this in a service release? @Eilon @rynowak

**Comment by joeaudette** [comment]
Update on this based on what @JamesNK posted, I got the idea to make my route constraint name match the route parameter by changing my foldertenant default route like this:

    routes.MapRoute(
                    name: "folderdefault",
                    template: "{sitefolder}/{controller}/{action}/{id?}",
                    defaults: new { controller = "Home", action = "Index" },
                    constraints: new { sitefolder = new cloudscribe.Core.Web.Components.SiteFolderRouteConstraint() });

After doing that and enabling endpointrouting, my breakpoint is hit in the Match method of my routeconstaint. But there is still a problem. The first request the RouteDirection parameter is IncomingRequest and the httpContext is not null so I can get my tenant. But on subsequent calls to the method when the routeDirection is "urlGeneration", httpContext is null and therefore I can't get my tenant and the urls for my links are not produced correctly, ie on the homepage of the folder tenant ie /two the login link should be /two/account/login but it is not because my route constraint although invoked did not match. I had to add a null check on the httpContext in my Match method which was not needed in 2.1, and if I disable endpointrouting the httpContext is never null and it works as expected.

**Comment by JamesNK** [comment]
That initial call to the constraint with a null `HttpContext` happens when the application starts up. Endpoint routing needs to check whether a route is valid for a controller, part of which is checking the route constraints. Because it happens outside the context of an HTTP request the constraints are called with a null `HttpContext`.

Your solution of passing when the context is null is fine. Multiple routes will match the endpoint and then at runtime the constraint will run when the context is present and the constraint will filter the request down to the correct endpoint. Your change will still be needed in ASP.NET Core 3.

**Comment by JamesNK** [comment]
@rynowak Idea: what do you think of limiting the constraints run against routes at startup to those that ship with ASP.NET? We know they are ok with a null context. Additional user created routes can still run at runtime and filter down the result to one endpoint.

**Comment by joeaudette** [comment]
@JamesNK I'm finding the null httpcontext is not just during application startup it is at runtime, every time I refresh the page for my folder tenant home page https://localhost:44399/two my route constraint is called once with routeDirection IncomingRequest and httpcontext is not null, then it gets called a number of times with routeDirection UrlGeneration but httpContext is null every time and therefore my constraint cannot match and urls on the page such as my login link are rendered incorrectly as https://localhost:44399/account/login when they should be https://localhost:44399/two/account.login

**Comment by joeaudette** [comment]
in fact my route constraint does not seem to be called at all during application startup.

**Comment by JamesNK** [comment]
Ok, I would have thought the route constraint would have been called at startup with a null context. I didn't read that you were getting it from link generation.

Link generation is quite separate from URL matching. How you are you generating the URLs? I assume it is the UrlHelper. Could you post code of how you call it?

**Comment by joeaudette** [comment]
For the login link I'm just using the taghelper like this:

    <a class="nav-link" 
            asp-controller="Account" 
            asp-action="Login" 
            asp-route-returnurl="@Context.Request.PathAndQuery()">Log in</a>

**Comment by JamesNK** [comment]
I'm afraid you've run into a new bug in endpoint routing. The HttpContext is not being passed to route constraints correctly in link generation. The workaround is to disable endpoint routing.

**Comment by joeaudette** [comment]
@JamesNK makes me happy to have reported a valid bug, the workaround is fine for me. Will this issue be used to track it or is there another issue to follow?

**Comment by JamesNK** [comment]
It is already fixed here for 3.0 - https://github.com/aspnet/AspNetCore/pull/6045

Because there is a simple workaround to this issue, disable endpoint routing in 2.2, the fix will likely wait for 3.0. If more people run into this issue then we'll re-prioritize it.

**Comment by joeaudette** [comment]
Fantastic! Feel free to close this issue.
